datalad/customremotes/ria_remote.py

```diff
@@ -0,0 +1,933 @@
+
+    def exists(self, path):
+        try:
+            self._run('test -e {}'.format(sh_quote(str(path))), check=True)
+            return True
+        except RemoteCommandFailedError:
+            return False
+
+    def in_archive(self, archive_path, file_path):
+
+        if not self.exists(archive_path):
+            return False
+
+        loc = str(file_path)
+        # query 7z for the specific object location, keeps the output
+        # lean, even for big archives
+        cmd = '7z l {} {}'.format(str(archive_path), loc)
+
+        # Note: Currently relies on file_path not showing up in case of failure
+        # including non-existent archive. If need be could be more sophisticated
+        # and called with check=True + catch RemoteCommandFailedError
+        out = self._run(cmd, no_output=False, check=False)
+
+        return loc in out
+
+    def get_from_archive(self, archive, src, dst, progress_cb):
+
+        # Note, that as we are in blocking mode, we can't easily fail on the actual get (that is 'cat').
+        # Therefore check beforehand.
+        if not self.exists(archive):
+            raise RIARemoteError("archive {arc} does not exist.".format(arc=archive))
+
+        # TODO: We probably need to check exitcode on stderr (via marker). If archive or content is missing we will
+        #       otherwise hang forever waiting for stdout to fill `size`
+
+        cmd = '7z x -so {} {}\n'.format(str(archive), str(src))
+        self.shell.stdin.write(cmd.encode())
+        self.shell.stdin.flush()
+
+        # TODO: - size needs double-check and some robustness
+        #       - can we assume src to be a posixpath?
+        #       - RF: Apart from the executed command this should be pretty much identical to self.get(), so move that
+        #         code into a common function
+
+        from os.path import basename
+        size = self._get_download_size_from_key(basename(str(src)))
+
+        with open(dst, 'wb') as target_file:
+            bytes_received = 0
+            while bytes_received < size:
+                c = self.shell.stdout.read1(1024)
+                if c:
+                    bytes_received += len(c)
+                    target_file.write(c)
+                    progress_cb(bytes_received)
+
+    def read_file(self, file_path):
+
+        cmd = "cat  {}".format(str(file_path))
+        try:
+            out = self._run(cmd, no_output=False, check=True)
+        except RemoteCommandFailedError:
+            raise RIARemoteError("Could not read {}".format(str(file_path)))
+
+        return out
+
+    def write_file(self, file_path, content, mode='w'):
+
+        if mode == 'w':
+            mode = ">"
+        elif mode == 'a':
+            mode = ">>"
+        else:
+            raise ValueError("Unknown mode '{}'".format(mode))
+        if not content.endswith('\n'):
+            content += '\n'
+
+        cmd = "echo \"{}\" {} {}".format(content, mode, str(file_path))
```

Comment:
Wouldn't this add newlines? `printf` might be a better choice.

Reply:
Added `-n` in PR #4203 already. Don't want to invest thoughts right now, since it is to be replaced by proper command abstraction in PR #4068 anyways.

Reply:
I think it is objective wrong here, and `echo -n` isn't well supported across platforms.

Reply:
Implemented in #4260